Only run one instance of daemon

jvff opened the pull request:
Before the RPC server is started, check if there is already another
instance of the daemon running. If there is, and it appears to be
responding correctly, don't start the daemon again.

Checklist for a PR:

* [x] Describe the change in **`CHANGELOG.md`**. Only applicable if the change has any impact for a user.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/mullvadvpn-app/87)
<!-- Reviewable:end -->

jvff commented:
Some open issues:

- [ ] should this change be in the changelog? If so, as a change or a fix?
- [ ] should there be a connection timeout? One of my tests was to start one daemon, send it a SIG_STOP and then start a second daemon, but the second instance got stuck (waiting for a reply, AFAIK). If timeout is necessary, should that be inside `WsIpcClient`?
- [ ] should the client connection code be moved to another crate (`mullvad-rpc`?), so that it is shared between the CLI and the backend?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L6rWd8fVi2V5e1tMjfK:b7egr2d)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s50iJ_V1ql4qzcPhA:-L6s50iJ_V1ql4qzcPhB:b-v6m3gn) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
> ```Rust
> 
> 
> pub fn no_other_instance_is_running() -> bool {
> ```

I prefer the positive version of names that checks for something. So rather check if another daemon is running rather than checking if no other is running. Check if other daemon is responding rather than not responding etc. Names usually becomes shorter that way and it becomes easier to reason about the boolean. Take this example: "Is the daemon not running: no" <- Takes longer to translate in at least my head than "is the daemon running: yes" :D

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6s5gLxspADg-M2ytFu:bswzscf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
> ```Rust
> 
> 
> pub fn no_other_instance_is_running() -> bool {
> ```

Could we maybe have this new feature in some other (new?) module? This module was from the start just a half-ugly way to place the code writing the address file somewhere.

We can keep the logic here for reading the content of any existing RPC address file, since that is highly related to knowing the hardcoded path and also closely related to writing the file. But I would prefer if the code for doing an RPC call was in it's own module or somewhere more related to that.

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6s4oqrwi3CNydnViCw:b930nsw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
> ```Rust
> pub fn write(rpc_address: &str, shared_secret: &str) -> Result<()> {
>     // Remove any existent RPC address file first
>     let _ = remove();
> ```

Since `open_file(...)` already has `.truncate(true)` there is no need in removing it before writing, just overwrite.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
See my trello comment about being in the changelog :)

A timeout might be reasonable indeed. If the daemon is localhost it should never take more than a few milliseconds to respond, so a timeout of 500ms or something would be good.

Yeah, the entire `WsIpcClient` code is a hack. It can't even support the authentication method we use for the GUI. We should really swap out the entire thing when we have time. When doing RPC over http we use a crate I wrote: https://github.com/mullvad/jsonrpc-client-rs/ but it does not yet support websockets. I have some local code laying around for trying to add that, but it's not done yet. It would be awesome if we got that done. Because it would make this easier and it would make the CLI support auth all in one go.

But that should probably not be done in this PR, no.

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L6s7OGzKORni2nqR0pn:brcxaew)*
<!-- Sent from Reviewable.io -->

jvff commented:
Understood. So for now leave it as is, with the known limitation about the timeout? Or should I hack a workaround, like performing the call in a separate thread and killing the thread if the timeout is reached?

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s50iJ_V1ql4qzcPhA:-L6upujl4FaCpLClnabH:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I prefer the positive version of names that checks for something. So rather check if another daemon is running rather than checking if no other is running. Check if other daemon is responding rather than not responding etc. Names usually becomes shorter that way and it becomes easier to reason about the boolean. Take this example: "Is the daemon not running: no" <- Takes longer to translate in at least my head than "is the daemon running: yes" :D
</blockquote></details>

Done.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6upyY4p31jBC59AS7K:b-xfd2q) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Could we maybe have this new feature in some other (new?) module? This module was from the start just a half-ugly way to place the code writing the address file somewhere.

We can keep the logic here for reading the content of any existing RPC address file, since that is highly related to knowing the hardcoded path and also closely related to writing the file. But I would prefer if the code for doing an RPC call was in it's own module or somewhere more related to that.
</blockquote></details>

Agreed. Should this go in a new module in this crate, or should it go in another crate and be shared with mullvad-cli?

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6uowLfleMP2LWOL9yn:bfc7vpz) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Since `open_file(...)` already has `.truncate(true)` there is no need in removing it before writing, just overwrite.
</blockquote></details>

I'll test this again, but I added it because when I created an existing file with permissions `user:user`, the daemon didn't overwrite it for some reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L6upeJbkTo6GOEAbGTw:b-1oo4pa)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6vLOa7hGJkrq2eZhxy:b-5setry) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Agreed. Should this go in a new module in this crate, or should it go in another crate and be shared with mullvad-cli?
</blockquote></details>

Just a new module in this crate. The reading file part is so tiny so better duplicate IMO.

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6vLFh1J7ElwcyEVbBM:b9flp1e) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I'll test this again, but I added it because when I created an existing file with permissions `user:user`, the daemon didn't overwrite it for some reason.
</blockquote></details>

It should be written as `root:root`, no? Since the daemon is executed as root. And the address file must be non-writable by unprivileged users.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6vWa5I1x1tr6WiYvil:b-rp2fv6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
> ```Rust
> 
> 
> pub fn other_instance_is_running() -> bool {
> ```

My two cents:

The preferred way of implementing this on Windows is by using a named mutex (i.e via `CreateMutexEx`) or named event (`CreateEvent`). 

On Linux the equivalent would be `sem_open()`, pthreads also offer shared memory mutex etc.. disclaimer: never did before.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6vm-B0bXFAOWOJqYAt:bpne3ma) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

My two cents:

The preferred way of implementing this on Windows is by using a named mutex (i.e via `CreateMutexEx`) or named event (`CreateEvent`). 

On Linux the equivalent would be `sem_open()`, pthreads also offer shared memory mutex etc.. disclaimer: never did before.
</blockquote></details>

But then we need a whole new way of communicating. Which feels out of scope for this tiny fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6wuH6Tqad7LX_8sxCa:bypzlhe) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It should be written as `root:root`, no? Since the daemon is executed as root. And the address file must be non-writable by unprivileged users.
</blockquote></details>

That's the normal behavior. But I tested the case where user ran it as non-root by mistake and then ran it as root. If the first daemon crashed, the second daemon doesn't start correctly. I believe this line prevents that from happening.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6wytOPdHIv-f7QA5ny:b-s8lgkm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

But then we need a whole new way of communicating. Which feels out of scope for this tiny fix.
</blockquote></details>

I took a look at named semaphores in Linux, and from what I understood, if a process dies the semaphore continues existing and isn't released (`sem_post`ed), so it would appear that the daemon is still running even though it has crashed. I tried to confirm this behavior here, but what I saw was worse: if the semaphore isn't unlinked, next time it is used (with `sem_trywait`) a segmentation fault occurs.

Hence I would prefer doing the RPC call to see if the daemon is "active" (might be running but not responding because it hang, for example).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6xIw57Iq5fw8904lw3:b-qqebi9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I took a look at named semaphores in Linux, and from what I understood, if a process dies the semaphore continues existing and isn't released (`sem_post`ed), so it would appear that the daemon is still running even though it has crashed. I tried to confirm this behavior here, but what I saw was worse: if the semaphore isn't unlinked, next time it is used (with `sem_trywait`) a segmentation fault occurs.

Hence I would prefer doing the RPC call to see if the daemon is "active" (might be running but not responding because it hang, for example).
</blockquote></details>

The idea behind is of course that it’s released when the process dies or quit. I know very little about how it’s done on Linux.

Not sure what Linus meant under new way of communicating.

Up to you

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6xZMLe6PZhr2buSlGV:bimpckn) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, pronebird (Andrei Mihailov) wrote…</i></summary><blockquote>

The idea behind is of course that it’s released when the process dies or quit. I know very little about how it’s done on Linux.

Not sure what Linus meant under new way of communicating.

Up to you
</blockquote></details>

@pronebird I meant that that type of solution requires things we have never used and likely new dependencies. While on the other hand we already have everything in place for just adding a simple RPC call, so it's more of a no-brainer.

Since it's a fairly low priority card the two options are 1) solving it in a simple way or 2) Don't solve it at all for now.
But it depends of course. If it turns out to be solvable in a nice way fast then that is of course the best. Just because we are time limited does not mean we have to take a bad solution if we know of a good one that can also be solved fast.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6xbg0-9EC7J1XhuOLR:be6vx2k) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

That's the normal behavior. But I tested the case where user ran it as non-root by mistake and then ran it as root. If the first daemon crashed, the second daemon doesn't start correctly. I believe this line prevents that from happening.
</blockquote></details>

I see. I tried it now. I got the behavior that if the file is owned by `user:user` and I then run as `root` it can still open, truncate and overwrite the file correctly. But the file is still owned by `user:user` afterwards, which would be a security problem indeed. Great find! Probably not a problem for real customers, since they never start the daemon manually, but still a hole we should fix.

With that in mind it might be reasonable to make sure the remove does not fail here, because then we can end up in that same security hole. Also, you have no real need for the chained error in `fn remove()` here. So if you just use `fs::remove_file(...)` and fail on all errors that are not `NotFound` I think we should be safe.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6yBgFfX8E5m1FpqKiD:batsoy1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Just a new module in this crate. The reading file part is so tiny so better duplicate IMO.
</blockquote></details>

I created an `rpc` crate and placed `rpc_info` inside it. Don't know if this is the best way to organize it; feedback is welcome.

---

*[mullvad-daemon/src/rpc_info.rs, line 44 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s4oqqb9jxz7D0bBWC:-L6y9usg9IOwQpoe_TFS:b2oqk7p) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L44)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I see. I tried it now. I got the behavior that if the file is owned by `user:user` and I then run as `root` it can still open, truncate and overwrite the file correctly. But the file is still owned by `user:user` afterwards, which would be a security problem indeed. Great find! Probably not a problem for real customers, since they never start the daemon manually, but still a hole we should fix.

With that in mind it might be reasonable to make sure the remove does not fail here, because then we can end up in that same security hole. Also, you have no real need for the chained error in `fn remove()` here. So if you just use `fs::remove_file(...)` and fail on all errors that are not `NotFound` I think we should be safe.
</blockquote></details>

I created a separate function to hopefully make the intention clear on what we're doing.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6yAMNLIse6Dmy4HLXQ:bvf3hpr) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

@pronebird I meant that that type of solution requires things we have never used and likely new dependencies. While on the other hand we already have everything in place for just adding a simple RPC call, so it's more of a no-brainer.

Since it's a fairly low priority card the two options are 1) solving it in a simple way or 2) Don't solve it at all for now.
But it depends of course. If it turns out to be solvable in a nice way fast then that is of course the best. Just because we are time limited does not mean we have to take a bad solution if we know of a good one that can also be solved fast.
</blockquote></details>

Perhaps a solution is to use file locks (https://gavv.github.io/blog/file-locks/). I haven't tested this, but it seems to be released when the process dies.

I think we should probably keep the RPC check, to avoid the case where the daemon got stuck and isn't responding, for example. In that case however we should probably kill the process.

I believe since this feature can grow into something more complicated, we should keep it as is and increment to it later. For example, lock the rpc_file and avoid a possible race condition if two daemons start up almost simultaneously. Then investigate killing the other process if it gets locked up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 5 files at r3.
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6zP2WOmGO0RHycd9eF:b-d2so0q) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I created an `rpc` crate and placed `rpc_info` inside it. Don't know if this is the best way to organize it; feedback is welcome.
</blockquote></details>

I have two arguments kind of against this organization. Now `rpc::rpc_info` has a redundant `rpc` in it. And not your fault, but I think `rpc_info` is a suboptimal name, "info" is such a generic word it's rarely helpful.

Also, our largest and most important RPC-related code in this entire crate is `management_interface`, which is not part of the `rpc` crate, making it a bit misrepresentative. And I don't think it should be under an `rpc` module either, it's very much a top level thing.

What are your thoughts on keeping a flat module structure and name `rpc_info` -> `rpc_address_file`? And add the RPC call code to a toplevel `rpc_client` maybe?

---

*[mullvad-daemon/src/rpc/mod.rs, line 15 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zQ5QjCEM_NygsWkjw:-L6zQ5QktRV0ZpZFRdkw:b8wgqyo) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc/mod.rs#L15)):*
> ```Rust
> /// other daemon has stopped.
> pub fn other_instance_is_running() -> bool {
>     rpc_info::exists() && other_daemon_responds()
> ```

`call_other_daemon` will read the RPC address, and it will fail if it does not exist. So the `exists()` check here is superfluous in its current state. But also `call_other_daemon` maybe does a bit too much since it both reads the address and does the call.

If you make it `check_other_daemon(address: &str) -> Result<()>` then you give it a more isolated responsibility and can reduce reading of the rpc address file to one location.

---

*[mullvad-daemon/src/rpc/mod.rs, line 21 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zRIH2RuGCg1YAu2tk:-L6zRIH2RuGCg1YAu2tl:b2nrvsd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc/mod.rs#L21)):*
> ```Rust
>     if let Err(message) = call_other_daemon() {
>         info!("{}; assuming it has stopped", message);
>         return false;
> ```

This method just converts a result to a boolean. I usually don't argue against splitting up stuff into small functions, but feels like this could be inlined with the above without losing anything :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Review status: 1 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r2](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6vWa5I1x1tr6WiYvik:-L6znVD0rnAgfAK8jmrA:b-tk2c1y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d5f064ef729f40ff199b998e2696de96451932bc/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Perhaps a solution is to use file locks (https://gavv.github.io/blog/file-locks/). I haven't tested this, but it seems to be released when the process dies.

I think we should probably keep the RPC check, to avoid the case where the daemon got stuck and isn't responding, for example. In that case however we should probably kill the process.

I believe since this feature can grow into something more complicated, we should keep it as is and increment to it later. For example, lock the rpc_file and avoid a possible race condition if two daemons start up almost simultaneously. Then investigate killing the other process if it gets locked up.
</blockquote></details>

@faern that's true, but it felt like that having a kernel object is more reliable way.

@jvff Good find. `flock` can probably work on Linux and macOS. Windows has `CreateFile` that accepts a `dwShareMode = 0` which seems to be working the same way but that has to be verified of course.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[mullvad-daemon/src/rpc_info.rs, line 35 at r1](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6s5gLxspADg-M2ytFt:-L6zqhXZn_TvqiCKefA9:b-wn8nbc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/364ddcc8d8d989568343eb3736e2ac0b58c9f356/mullvad-daemon/src/rpc_info.rs#L35)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I have two arguments kind of against this organization. Now `rpc::rpc_info` has a redundant `rpc` in it. And not your fault, but I think `rpc_info` is a suboptimal name, "info" is such a generic word it's rarely helpful.

Also, our largest and most important RPC-related code in this entire crate is `management_interface`, which is not part of the `rpc` crate, making it a bit misrepresentative. And I don't think it should be under an `rpc` module either, it's very much a top level thing.

What are your thoughts on keeping a flat module structure and name `rpc_info` -> `rpc_address_file`? And add the RPC call code to a toplevel `rpc_client` maybe?
</blockquote></details>

I agree on `rpc_address_file`. I'm not sure if there's an alternative to `rpc_client` that's clearer, because currently the only use case is checking if another instance is running. Perhaps something like `rpc_uniqueness_check`, `daemon_uniqueness`, or `unique_instance`?

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zQ5QjCEM_NygsWkjw:-L6zwR5jzzWCW-oeHJwe:b-ducshs) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

`call_other_daemon` will read the RPC address, and it will fail if it does not exist. So the `exists()` check here is superfluous in its current state. But also `call_other_daemon` maybe does a bit too much since it both reads the address and does the call.

If you make it `check_other_daemon(address: &str) -> Result<()>` then you give it a more isolated responsibility and can reduce reading of the rpc address file to one location.
</blockquote></details>

Merged the two functions. I considered changing the `other_instance_is_running` function body into:

    call_other_daemon()
        .map_err(|message| info!("{}; assuming it has stopped", message))
        .is_ok()

but although it looks cleaner to me, I'm uncertain if it becomes less obvious what's happening. 

---

*[mullvad-daemon/src/rpc/mod.rs, line 21 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zRIH2RuGCg1YAu2tk:-L6zuEVpaMa5j3EId9Mu:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc/mod.rs#L21)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This method just converts a result to a boolean. I usually don't argue against splitting up stuff into small functions, but feels like this could be inlined with the above without losing anything :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 5 files at r4.
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 58 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zziOqS5Oe-P9CnT3V:-L6zziOqS5Oe-P9CnT3W:b-q6mx6r) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L58)):*
> ```Rust
> /// Removes the created RPC file.
> ///
> /// Fails if the file doesn't exist.
> ```

It does not just fail on non existent files. It fails on any error during removal, such as insufficient permissions or anything else.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 67 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7--btyEKi6jg2bYall:-L7--btyEKi6jg2bYalm:b-8jb93v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L67)):*
> ```Rust
> ///
> /// Doesn't fail if the file doesn't exist.
> fn remove_previous() -> Result<()> {
> ```

I don't think the name fully convey what it does. My first thought when I read it was "oh, what previous file, do we have two paths?". When in reality it removes but silences the missing error. It's about naming after how it's used (remove previous) vs naming what it does (remove but ignore missing).

I think it would be fine for the normal `remove()` to also just silence the missing error, would it not? If we at shutdown try to remove the address file and it's already gone that is not a blocker or critical error. So I think we would be fine with a single `remove` function that always silences the `NotFound` error.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 69 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7--7bvMSUI9XO3guna:-L7--7bvMSUI9XO3gunb:b-4aaxyr) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L69)):*
> ```Rust
> fn remove_previous() -> Result<()> {
>     match fs::remove_file(RPC_ADDRESS_FILE_PATH.as_path()) {
>         Err(error) => {
> ```

Completely up to you, mostly a style thing. But I usually find that `if let ... = ...` falls out nicer when you only care about one of the pattern cases. Not saying you have to change, it's just a tip :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 58 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zziOqS5Oe-P9CnT3V:-L7-2lNWOMS_M1DfFwvi:b-2lxyzi) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L58)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It does not just fail on non existent files. It fails on any error during removal, such as insufficient permissions or anything else.
</blockquote></details>

I tried to distinguish the two functions, but I agree that doing so as I did it made the documentation unclear. Shouldn't be a problem since I merged the two.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 67 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7--btyEKi6jg2bYall:-L7-30CU300BlDVk9Qh-:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L67)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I don't think the name fully convey what it does. My first thought when I read it was "oh, what previous file, do we have two paths?". When in reality it removes but silences the missing error. It's about naming after how it's used (remove previous) vs naming what it does (remove but ignore missing).

I think it would be fine for the normal `remove()` to also just silence the missing error, would it not? If we at shutdown try to remove the address file and it's already gone that is not a blocker or critical error. So I think we would be fine with a single `remove` function that always silences the `NotFound` error.
</blockquote></details>

Done.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 69 at r4](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7--7bvMSUI9XO3guna:-L7-369oJxYqkOhIkDMn:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9254bf805643cdbd69bd8a3a792fb19b9041abad/mullvad-daemon/src/rpc_address_file.rs#L69)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Completely up to you, mostly a style thing. But I usually find that `if let ... = ...` falls out nicer when you only care about one of the pattern cases. Not saying you have to change, it's just a tip :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 7 files at r3.
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 8 at r5](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7-Eaipl6kukJG89K3Y:-L7-EaiqzHMsx8l3dmNc:bqvsksz) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/336e9c06a2ef3b21655f0db2a7ae0d557c16dce0/mullvad-daemon/src/rpc_address_file.rs#L8)):*
> ```Rust
>     errors {
>         WriteFailed(path: PathBuf) {
>             description("Failed to write RCP connection info to file")
> ```

Typo: RCP. Not introduced by you, but I spotted it now :)

---

*[mullvad-daemon/src/rpc_address_file.rs, line 61 at r5](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7-PQ4UdjZSGqWcx2C_:-L7-PQ4UdjZSGqWcx2Ca:b-1z8vz4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/336e9c06a2ef3b21655f0db2a7ae0d557c16dce0/mullvad-daemon/src/rpc_address_file.rs#L61)):*
> ```Rust
>         if error.kind() == io::ErrorKind::NotFound {
>             // No previously existing file
>             return Ok(());
> ```

I don't know about you, but I like the expression oriented way and avoiding `return`. If you place the last lonely `Ok(())` in an `else {}` branch to the initial `if let ...` you can get rid of both `return`s and have the entire function be one expression :)

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zQ5QjCEM_NygsWkjw:-L7-QCxXUSJmxZBnhdAA:bzc5ay6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Merged the two functions. I considered changing the `other_instance_is_running` function body into:

    call_other_daemon()
        .map_err(|message| info!("{}; assuming it has stopped", message))
        .is_ok()

but although it looks cleaner to me, I'm uncertain if it becomes less obvious what's happening. 
</blockquote></details>

I usually like chaining operators indeed. But I'm not a huge fan of using `map_err`/`map` for their side effects. You are just mapping the error to `()` here just to get the side effect. So since you are after the side effects I prefer the `if let` construct. However, you can remove `return` in both instances.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_address_file.rs, line 8 at r5](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7-Eaipl6kukJG89K3Y:-L7-VLt2P_52DKy3Vwzl:busawlc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/336e9c06a2ef3b21655f0db2a7ae0d557c16dce0/mullvad-daemon/src/rpc_address_file.rs#L8)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Typo: RCP. Not introduced by you, but I spotted it now :)
</blockquote></details>

Fixed :)

---

*[mullvad-daemon/src/rpc_address_file.rs, line 61 at r5](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L7-PQ4UdjZSGqWcx2C_:-L7-VOaX2IuRsWrCRaHX:b-phytx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/336e9c06a2ef3b21655f0db2a7ae0d557c16dce0/mullvad-daemon/src/rpc_address_file.rs#L61)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I don't know about you, but I like the expression oriented way and avoiding `return`. If you place the last lonely `Ok(())` in an `else {}` branch to the initial `if let ...` you can get rid of both `return`s and have the entire function be one expression :)
</blockquote></details>

Agreed, it looks better now!

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r3](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L6zQ5QjCEM_NygsWkjw:-L7-Vc15L9Xo__kLMx03:b-8pmh0y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/da96116c22cd3a9dc65862103358e355a90873c2/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I usually like chaining operators indeed. But I'm not a huge fan of using `map_err`/`map` for their side effects. You are just mapping the error to `()` here just to get the side effect. So since you are after the side effects I prefer the `if let` construct. However, you can remove `return` in both instances.
</blockquote></details>

I hadn't thought about the `map`/`map_err` misuse for side-effects, but it makes sense. It's not mapping, it's acting upon. Would be cool to have a `.log_err` trait extension method instead :)

I've kept it the way it was and removed the unnecessary `return`s. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 5 files at r4, 1 of 2 files at r6.
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 9 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70MltsgcH9eO-Pzt13:-L70Mltt0qPpswXxYJq-:b1td5m4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L9)):*
> ```Rust
> 
> 
> /// Verifies if there is another instance of the daemon running.
> ```

s/verifies/checks/ ?

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 13 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70Nf4fu20VrlKINmnN:-L70Nf4fu20VrlKINmnO:bqtwuw7) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L13)):*
> ```Rust
> /// Tries to connect to another daemon and perform a simple RPC call. If it fails, assumes the
> /// other daemon has stopped.
> pub fn other_instance_is_running() -> bool {
> ```

A bit nitpicking: Functions returning booleans usually start with `is_` if they contain "is" at all. How about `is_other_instance_running`?

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70N4dhEbShLDW79AqW:-L70N4diwmSG6ikMQOCi:b2cnxv0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
> ```Rust
> pub fn other_instance_is_running() -> bool {
>     if let Err(message) = call_other_daemon() {
>         info!("{}; assuming it has stopped", message);
> ```

This will happen 100% of all daemon starts. Even when there is no address file present. This check is more to cover an edge case. So I don't feel we want this logged all the time.

Which is why I suggested loading the address first and only if it exists try to `call_other_daemon` and log the result.

Or keep the structure as is but remove this logging and name `call_other_daemon` -> `check_other_instance_running` or something, since it does more than just the call itself.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 2 files at r7.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70N4dhEbShLDW79AqW:-L73T2Rvsop5Th6IYxjh:bhpi5qx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I think now I understood what you meant. I changed it to load the address first and then pass it to `call_other_instance` (swapped daemon with instance in the name to make things consistent). Hopefully this is better?

Also, I created a `check_result` function to handle the log printing. Do you think it helps with clarity? (Last "new idea" for this PR, I promise).
</blockquote></details>

I like the structure around reading the RPC file and then calling.

`check_result` is basically a `try!` implementation with logging almost. And you have to call it twice. The responsibility would be much clearer split up if `call_other_instance` only called another instance and returned the result of that call. Having `call_other_instance` log "assuming it has stopped" sounds like someone elses responsibility to me.

`is_another_instance_running` sounds like the perfect place to log to me. It is responsible for checking if another instance is running, and can thus log if it determines one does not. It would not bloat this function by much to inline the code in `check_result` here, just another `match` or `if let`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 5 files at r4, 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70N4dhEbShLDW79AqW:-L74BwchKze7S1PHRspG:bk340vt) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I like the structure around reading the RPC file and then calling.

`check_result` is basically a `try!` implementation with logging almost. And you have to call it twice. The responsibility would be much clearer split up if `call_other_instance` only called another instance and returned the result of that call. Having `call_other_instance` log "assuming it has stopped" sounds like someone elses responsibility to me.

`is_another_instance_running` sounds like the perfect place to log to me. It is responsible for checking if another instance is running, and can thus log if it determines one does not. It would not bloat this function by much to inline the code in `check_result` here, just another `match` or `if let`.
</blockquote></details>

Done. I decided to use `match` instead of `if let` for two reasons, but I'm not sure if they're valid reasons or just style quirks:

1. The returned values were in an unusual order when reading (`false`, `true`, `false`);
2. The `if let` had to bind the `Err` instead of the `Ok` result, which is the opposite of the general usage trend.

What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

jvff commented:
I think there's a problem with this PR in its current form. Suppose two daemons start up almost simultaneously. Before the check for the other daemon is performed, they both might try to update the relay list. If that happens, I'm not sure if data corruption can happen.

To prevent this from happening, I moved the check to early in the `Daemon::new()` constructor. Now the problem is the check may pass even though both are starting (because the check happens before either of them actually starts the RPC server).

I'm trying to think of a simple solution, at least temporarily, but so far all I've got is to create the file sooner with no contents, and then wait up to a minute (more or less?) while waiting for the file to be populated with the address, and then attempting the RPC call.

Any alternatives or thoughts on the issue?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L74D_mQ_OuLhqFfw2gq:bpxpicg)*
<!-- Sent from Reviewable.io -->

faern commented:
It's really hard to solve this in an non-fallible way. Reading and writing files almost can't be atomic operations. So not much to do about it. Or we can get fancy and try different things. But this is quite a low priority issue and I don't see the need. This is still a lot better than before this PR.

---

Reviewed 1 of 5 files at r4, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[mullvad-daemon/src/rpc_uniqueness_check.rs, line 15 at r6](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-L70N4dhEbShLDW79AqW:-L74QzacM8wJA0xLA4RB:bubgw6d) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/bf5d1f71c7d842a3df26da896097415cd69a2b6d/mullvad-daemon/src/rpc_uniqueness_check.rs#L15)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Done. I decided to use `match` instead of `if let` for two reasons, but I'm not sure if they're valid reasons or just style quirks:

1. The returned values were in an unusual order when reading (`false`, `true`, `false`);
2. The `if let` had to bind the `Err` instead of the `Ok` result, which is the opposite of the general usage trend.

What do you think?
</blockquote></details>

I quite often do `if let Err(e) = ...`. I don't find it unusual. Quite often an operation return `Result<()>` and I don't care about the value of the success, but I want to "catch" any error.

But this is of course fine as it is. Now we are down to style and just details. No need to overthink this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L74QF7t4JjTSgZQUbQi:b-w0frvi)*
<!-- Sent from Reviewable.io -->

jvff commented:
Ok, I'll leave it to a new future PR then :)

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87#-:-L74TADliVWmM9kgFMJV:b-kvr3o)*
<!-- Sent from Reviewable.io -->

eriklarko commented:
Reviewed 2 of 7 files at r3, 1 of 4 files at r6, 1 of 3 files at r7, 1 of 2 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 2 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

pronebird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/mullvadvpn-app/87#-:-L79A5nP5Y8Y7-IzVJYu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pronebird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/mullvadvpn-app/87)*
<!-- Sent from Reviewable.io -->
